File: python/cote/LTCD/팰린드롬_연결_리스트.py
from typing import *
from collections import deque
import random

TEST_CASES = 10
N = 100000

class Node :
	
	def __init__(self, value, nextNode=None) :
		self.value = value
		self.nextNode = nextNode

def solve1(head : Node) -> bool :

	q = deque()

	if not head : 
		return True

	node = head
	while node is not None :
		q.append(node.value)
		node = node.nextNode

	while len(q) > 1 :
		if q.popleft() != q.pop() :
			return False

	return True

## Runner 알고리즘
def solve2(head : Node) -> bool :

	rev = None
	slow = fast = head

	while fast and fast.nextNode :
		fast = fast.nextNode.nextNode	
		rev, rev.nextNode, slow = slow, rev, slow.nextNode
		
	if fast :
		slow = slow.nextNode

	while rev and rev.value == slow.value :
		slow, rev = slow.nextNode, rev.nextNode 

	return not rev

if __name__ == "__main__" :

	with open("data/팰린드롬_연결_리스트.txt") as fout :

		TEST_CASES = int(fout.readline()) 

		for _ in range(TEST_CASES) :
			nodes = [ Node(int(token)) for token in fout.readline().strip().split(" ")]
			for i in range(0, len(nodes)-1) :
				nodes[i].nextNode = nodes[i+1]

			print(solve1(nodes[0]), solve2(nodes[0]))
				

File: kotlin/src/main/kotlin/com/example/cote/LTCD/팰린드롬_연결_리스트.kt
package com.example.cote.LTCD

import java.io.*
import java.nio.file.*

object 팰린드롬_연결_리스트 {
    class Node(val value: Int) {
        var next: Node? = null
    }

    fun solve(head: Node): Boolean {
        val queue = ArrayDeque<Int>()

        var cur: Node? = head
        while (cur != null) {
            queue.addLast(cur.value)
            cur = cur.next
        }

        while (queue.size > 1) {
            if (queue.removeFirst() != queue.removeLast()) {
                return false
            }
        }

        return true
    }

    // Runner 알고리즘
    fun solve2(head: Node): Boolean {

        var fast: Node? = head
        var slow: Node? = head
        var rev: Node? = null

        while (fast?.next != null) {
            fast = fast.next?.next
            var slowNext = slow!!.next
            var revPrev = rev
            rev = slow
            rev.next = revPrev
            slow = slowNext
        }

        if (fast != null) {
            slow = slow?.next
        }

        while (rev != null && rev.value == slow!!.value) {
            rev = rev.next
            slow = slow.next
        }

        return rev == null
    }

    fun main() {
        val p =
            Paths.get(System.getProperty("user.dir") + "/data/팰린드롬_연결_리스트.txt")
        val reader = Files.newBufferedReader(p)
        repeat(reader.readLine().toInt()) {
            val nodes =
                reader
                    .readLine()
                    .trim('\n')
                    .split(" ")
                    .map(String::toInt)
                    .map(::Node)

            repeat(nodes.size - 1) { nodes[it].next = nodes[it + 1] }

            println("" + solve(nodes[0]) + " " + solve2(nodes[0]))
        }
    }
}

fun main() {
    팰린드롬_연결_리스트.main()
}
